library/Icingadb/ProvidedHook/Reporting/Common/SlaTimeline.php

```diff
@@ -0,0 +1,204 @@
+<?php
+
+/* Icinga DB Web | (c) 2023 Icinga GmbH | GPLv2 */
+
+namespace Icinga\Module\Icingadb\ProvidedHook\Reporting\Common;
+
+use Countable;
+use DateTimeInterface;
+
+/**
+ * Represents a single sla timeline for a given timeframe
+ */
+class SlaTimeline implements Countable
+{
+    /** @var DateTimeInterface Start time of the generated report timeframe */
+    protected $start;
+
+    /** @var DateTimeInterface End time of the generated report timeframe */
+    protected $end;
+
+    /** @var array Sla event time history within a given timeframe */
+    protected $time = [];
+
+    /** @var array Sla event type within a given timeframe */
+    protected $event = [];
+
+    /** @var array Sla history hard_state within a given timeframe */
+    protected $state = [];
+
+    /** @var array Sla history previous hard_state within a given timeframe */
+    protected $previousState = [];
+
+    /** @var int Sum of the problem time of this timeline */
+    protected $problemTime = 0;
+
+    /** @var int Total time of this timeline */
+    protected $totalTime = 0;
+
+    /** @var int The initial hard state of this timeline */
+    protected $initialHardState = 0;
+
+    public function __construct(DateTimeInterface $start, DateTimeInterface $end)
+    {
+        $this->start = clone $start;
+        $this->end = clone $end;
+    }
+
+    /**
+     * Set the initial hard state of this timeline
+     *
+     * @param int $state
+     *
+     * @return $this
+     */
+    public function setInitialHardState(int $state): self
+    {
+        $this->initialHardState = $state;
+
+        return $this;
+    }
+
+    /**
+     * Get the calculated SLA result of this timeline
+     *
+     * @return ?float
+     */
+    public function getResult(): ?float
+    {
+        $problemTime = 0;
+        $activeDowntimes = 0;
+        $lastEventTime = (int) $this->start->format('Uv');
+        $totalTime = (int) $this->end->format('Uv') - $lastEventTime;
+
+        $lastHardState = $this->initialHardState;
+        $count = $this->count();
+        for ($i = 0; $i < $count; ++$i) {
+            $event = $this->event[$i];
+            $time = $this->time[$i];
+            $state = $this->state[$i];
+
+            if ($this->previousState[$i] === 99 || ($lastHardState === 99 && $event !== 'state_change')) {
```

Comment:
This condition is changed to compared to the current one: https://github.com/Icinga/icingadb/blob/4ed4db3ab642622bb917797a8207589ce5f89f07/schema/mysql/schema.sql#L139

I think this is the change that's actually supposed to address https://github.com/Icinga/icingadb/issues/558. If we want to mix this in here, I'd try change the order of actions in the loop, the current condition was my suggestion for a quick test to see if this would bring an improvement. The general idea was to check if `$lastHardState` would be 99 at the end of the iteration. So it could make sense to just move this branch of the conditional after the update to `$lastHardState` (the update of `$lastEventTime` would have to be moved behind that then).

Reply:
> So it could make sense to just move this branch of the conditional after the update to `$lastHardState` (the update of `$lastEventTime` would have to be moved behind that then).

`$lastHardState` is only updated when the current row event type is a `state_change`. How would this make any difference compared the current implementation?

Reply:
It should do exactly the same. My hope is that the change makes the code easier to understand.